rvzr/model_dynamorio/backend/speculator_abc.cpp

```diff
@@ -146,17 +147,44 @@ pc_t SpeculatorABC::handle_instruction(instr_obs_t instr, dr_mcontext_t *mc, voi
     return 0;
 }
 
-void SpeculatorABC::handle_mem_access(bool is_write, void *address, uint64_t /*size*/)
+bool SpeculatorABC::handle_mem_access(bool is_write, void *address, uint64_t size)
 {
     if (not in_speculation)
         return;
 
     // record changes made to the memory
     if (is_write) {
-        // NOTE: it would be more correct to use dr_safe_read here to avoid faults;
-        // However, this code is on the hot path, and dr_safe_read is slow,
-        // so we just accept that horrible things may happen. Oh well.
-        const uint64_t val = *(uint64_t *)address;
-        store_log.push_back({.addr = (uint64_t)address, .val = val, .nesting_level = nesting});
+        auto cur_address = (uint64_t)address;
```

Comment:
The code will be slightly easier to read if you rewrite it to `if not is_write -> return`